File: Algorithm/Stack/Solution/스택 - boj/BOJ10828.c
#include <stdio.h>
#include <string.h>

int main(void)
{
    int N;
    scanf("%d ", &N);

    int stack[10000];
    int start = 0;
    int end = 0;

    for (int c = 0; c < N; ++c)
    {
        char command[7];
        scanf("%s", command);

        if (strcmp(command, "push") == 0)
        {
            scanf("%d ", stack + end);
            end++;
        }
        else if (strcmp(command, "pop") == 0)
        {
            if (start == end)
            {
                printf("-1\n");
            }
            else
            {
                printf("%d\n", stack[end - 1]);
                end--;
            }
        }
        else if (strcmp(command, "size") == 0)
        {
            printf("%d\n", end - start);
        }
        else if (strcmp(command, "empty") == 0)
        {
            printf("%d\n", end - start == 0);
        }
        else if (strcmp(command, "top") == 0)
        {
            if (start == end)
            {
                printf("-1\n");
            }
            else
            {
                printf("%d\n", stack[end - 1]);
            }
        }
        else
        {
            printf("??");
        }
    }

    return 0;
}

File: Algorithm/Stack/Solution/스택 - boj/BOJ10828.py
# python list의 method 사용 금지(미리 원하는 크기로 초기화 한 후 인덱스로 접근해야 함)
# len 함수 사용 금지
# deque 사용 금지

import sys

input = sys.stdin.readline
N = int(input())
stack = [0 for i in range(N)]
size = 0

for i in range(N):
    command = input().strip()

    if command.startswith('push'):
        stack[size] = int(str(command[5:]))
        size += 1
    elif command == 'pop':
        if size == 0:
            print(-1)
        else:
            print(stack[size-1])
            size -= 1
    elif command == 'size':
        print(size)
    elif command == 'empty':
        print(1 if size == 0 else 0)
    elif command == 'top':
        print(stack[size-1] if size != 0 else -1)
